### PR TITLE
[Master]-Bug 641503 Financial Reporting - Duplicate Rows in Financial Report Excel Export After Adding GL Budget Dimension

### PR DESCRIPTION
**Pull request by v-praghav**
[AB#641503](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/641503)

## Linked work
[AB#641503](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/641503)

Fixes #

## How I validated this

- [x] I read the full diff and it contains only changes I intended.
- [x] I built the affected app(s) locally with no new analyzer warnings.
- [x] I ran the change in Business Central and confirmed it behaves as expected.
- [x] I added or updated tests for the new behavior, or explained below why none are needed.

**What I tested and the outcome** *(required — be specific: scenarios, commands, screenshots for UI changes)*

### CODE REVIEW

**Comment by v-praghav** [comment]
@microsoft-github-policy-service agree company="Microsoft"

File: src/Layers/W1/Tests/ERM/ERMAccountScheduleII.Codeunit.al [hunk 1]
```diff
@@ -3431,6 +3484,24 @@
...
+        ValueFound: Boolean;
+        CellValue: Text[250];
+        OccurrenceCount: Integer;
+    begin
+        OccurrenceCount := 0;
+        for RowNo := 1 to 200 do
+            for ColumnNo := 1 to 20 do begin
+                CellValue := LibraryReportValidation.GetValueAt(ValueFound, RowNo, ColumnNo);
+                if ValueFound and (StrPos(CellValue, TextValue) > 0) then
+                    OccurrenceCount += 1;
```
[the remark this answers is not in the document]
Comment (v-praghav) [reply src/Layers/W1/Tests/ERM/ERMAccountScheduleII.Codeunit.al:3499]: Not Needed

**Review by v-ankitgoyal — DISMISSED** [review]

**Review by AndersLarsenMicrosoft — DISMISSED** [review]

**Review by v-ankitgoyal — DISMISSED** [review]

File: src/Layers/W1/Tests/ERM/ERMAccountScheduleII.Codeunit.al [hunk 2]
```diff
@@ -3431,6 +3484,24 @@
         FinRepStatus.Insert(true);
     end;
 
+    local procedure VerifyTextValueAppearsOnce(TextValue: Text[20])
+    var
+        RowNo: Integer;
```
[the remark this answers is not in the document]
Comment (v-praghav) [reply src/Layers/W1/Tests/ERM/ERMAccountScheduleII.Codeunit.al:3488]: Corrected

**Comment by alexei-dobriansky** [comment]
## Agentic PR Review - Round 1

# **Recommendation: Request Changes**

<details>
<summary><strong>What this PR does</strong></summary>

This PR changes Report 29 in W1 and NA to write blank Excel cells around the filter, currency, and column header area. It also adds W1, NA, and CZ tests that check dimension filter labels and that `Currency` appears once.

The code is trying to fix stale cells left in an existing Excel layout when the filter rows move after adding a G/L Budget or dimension filter. That is the right area, and no event publisher or public event signature was changed. However, the added tests currently break the test codeunit and do not prove the real Default Excel Layout path, G/L Budget filter path, row count, or exported amounts.
</details>

<details>
<summary><strong>Suggestions</strong></summary>

**S1** - *Duplicate test names break the build*
The new `ExportAccScheduleToExcelWithDimFilter` test has the same name as an existing test in the W1, NA, and CZ test codeunits. Rename or merge it before merge, because this makes the app fail to build.

**S2** - *Regression test misses the reported path*
The new test runs a fresh Report 29 export, but the bug is in Open in Excel using an imported Default Excel Layout with a G/L Budget filter. Add a test through the existing template path and assert no duplicate rows or labels, the expected row count, and the expected amount.

**S3** - *Blank currency can insert a duplicate cell*
`InsertBlankRowAndFillCells(RowNo, 3, false, true)` runs even if `EnterFilterInCell` did not add the currency row. If `GLSetup."LCY Code"` or the additional reporting currency is blank, it can insert cell 3 in the blank row that was already inserted. Guard this call or move it inside the branch that actually writes the currency row.
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** Report 29 feeds Financial Reporting Excel output, including customer layouts. A wrong fix can leave stale cells in templates, hide duplicate rows, or change exported financial values. The change is narrow and does not change public events, but it touches a financially sensitive export path and current PR checks are failing.

**Necessity:** The bug is valid and important. AB#641503 is a Bug, and the linked work item fields describe a customer closing/reporting issue with duplicated Excel rows after using dimension and G/L Budget filters. No `[AI-REPRO]` comment was found on the linked work item, but the repro fields and related parent bugs are concrete enough to justify a targeted fix.
</details>

---
> [AI-PR-REVIEW] version=1 system=github pr=9129 round=1 by=alexei-dobriansky at=2026-07-08 lastSha=9fd372499d9b1daec6f8d5a9e3172e0cafa6715a suggestions=S1,S2,S3

**Comment by alexei-dobriansky** [comment]
@v-praghav please review the suggestions

**Review by AndersLarsenMicrosoft — DISMISSED** [review]

**Comment by v-praghav** [comment]
Hi @alexei-dobriansky ,
I habve checked the Suggestions,  below is my comments
Suggettion 1 Reply: Corrected the name of procedure.
Suggestion 2 Reply:  I reviewed the scenario and believe the existing automation already covers it. The test runs "Export Acc. Sched. to Excel" and validates the relevant fields, so no automation changes seem necessary.
Suggestion 3 Reply: The blank row is intentional. If we don't insert a blank entry in the Excel Buffer, Excel generation can copy the data into a non-existing cell in the same column. We need this blank row before the currency insertion. Previously, that row was skipped, which is what caused the reported issue.

**Review by v-ankitgoyal — APPROVED** [review]

**Comment by alexei-dobriansky** [comment]
## Agentic PR Review - Round 2

# **Recommendation: Request Changes**

<details>
<summary><strong>What this PR does</strong></summary>

Since round 1, the PR only renames the duplicate new test in W1, NA, and CZ. That fixes the build issue from the duplicate procedure name.

The Report 29 code and the regression coverage are otherwise unchanged. The test still uses a fresh Report 29 export, not an imported Default Excel Layout with a G/L Budget filter, and the blank-cell code can still write the same blank cell twice when no currency row is written.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|----|-------|--------|-----------------|
| S1 | Duplicate test names break the build | Addressed | The new test was renamed to `ExportAccScheduleToExcelWithDimFilterPostedGLEntries` in W1, NA, and CZ. |
| S2 | Regression test misses the reported path | Disputed | The reply says the existing export test covers the scenario and no automation change is needed. The test still does not use an imported Default Excel Layout or a G/L Budget filter. |
| S3 | Blank currency can insert a duplicate cell | Disputed | The reply says the blank row is intentional. The single blank cell after the currency call still runs even when no currency row was written. |
</details>

<details>
<summary><strong>New observations (commits since round 1)</strong></summary>

*None - the only round-2 code change is the test rename. The remaining concerns are the disputed round-1 suggestions above.*
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The extra round-2 commit is low risk by itself, but it does not reduce the main risk. Report 29 writes Financial Reporting Excel output, including customer templates. If the blank cells are wrong, existing layouts can still show stale labels, duplicated cells, or wrong financial report rows.

**Necessity:** AB#641503 is a valid Bug with a concrete customer scenario: imported Default Excel Layout, a dimension filter, and a G/L Budget filter. A targeted Report 29 fix is needed, but it should be backed by a test that follows that path and the blank currency-row edge case should be made safe.
</details>

---
> [AI-PR-REVIEW] version=1 system=github pr=9129 round=2 by=alexei-dobriansky at=2026-07-15T05:39:50Z lastSha=dfe3b7cfec25554ccb4bf51e35d07a090ef4bfc5 suggestions=S1:addressed,S2:disputed,S3:disputed parentRound=1

**Review by AndersLarsenMicrosoft — APPROVED** [review]
